### ISSUE TITLE
Mypy CI job failing after numpy 2.0 release

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#20982]
Failing consistently. Failures from [this CI log](https://github.com/scipy/scipy/actions/runs/9540215732/job/26291647416?pr=20971):

```
💻  mypy.api.run --config-file /home/runner/work/scipy/scipy/mypy.ini scipy
scipy/_lib/_array_api.py:453: error: Incompatible return value type (got "None", expected Module)  [return-value]
scipy/_lib/_util.py:31: error: Module "numpy" has no attribute "AxisError"  [attr-defined]
scipy/_lib/_util.py:31: error: Name "AxisError" already defined on line 21  [no-redef]
scipy/_lib/_util.py:31: error: Module "numpy" has no attribute "ComplexWarning"  [attr-defined]
scipy/_lib/_util.py:31: error: Name "ComplexWarning" already defined on line 22  [no-redef]
scipy/_lib/_util.py:31: error: Module "numpy" has no attribute "VisibleDeprecationWarning"  [attr-defined]
scipy/_lib/_util.py:31: error: Name "VisibleDeprecationWarning" already defined on line 23  [no-redef]
Found 7 errors in 2 files (checked 889 source files)
```

I'll have a look at fixing this up.

**Comment by ilayn** [comment]
I thought #20979 fixed this

**Comment by rgommers** [comment]
Ah thanks, it does indeed - missed that PR. I got notifications after that was merged, but those were from CI on the 1.14.x branch. So adding the backport label to gh-20979.
